### DISCUSSION TITLE
Passando mais de um parâmetro da query no endpoint

### DISCUSSION THREAD

**Discussion by alvaroflp** [HashLoad/horse#309] [Q&A] [answered]
Como eu posso passar mais de um parâmetro no endpoint da API que estou construindo? Endpoints com somente um parametro funciona normalmente, porém quando eu faço com mais de um caio no except. 

`Controller.Price`

unit Controller.Price;

interface

uses
  Horse,
  System.JSON,
  System.SysUtils,
  Model.Price,
  FireDAC.Comp.Client,
  Data.DB,
  DataSet.Serialize;

procedure Endpoint;

implementation

{ procedure request / response without params }
procedure ListPrice(Req: THorseRequest; Res: THorseResponse; Next: TProc);
var
  Price: TPrice;
  qry: TFDQuery;
  erro: string;
  PriceArray: TJSONArray;
begin
  try
    Price := TPrice.Create;
  except
    Res.Send('Erro ao conectar ao banco de dados')
      .Status(500);
    exit;
  end;

  try
    qry := Price.ListPrice('', erro);
    PriceArray := qry.ToJSONArray();
    Res.Send<TJSONArray>(PriceArray);
  finally
    qry.Free;
    Price.Free;
  end;
end;

{ procedure request / response with params }
procedure ListPriceParams(Req: THorseRequest; Res: THorseResponse;
  Next: TProc);
var
  Price: TPrice;
  objPrice: TJSONObject;
  qry: TFDQuery;
  erro: string;
  PriceArray: TJSONArray;
begin
  try
    Price := TPrice.Create;
    Price.CODAUXILIAR := Req.Params['codauxiliar'].ToUpper;
    Price.CODFILIAL := Req.Params['codfilial'].ToUpper;
    Price.NUMREGIAO := Req.Params['numregiao'].ToUpper;
  except
    Res.Send('Erro ao conectar ao banco de dados')
      .Status(500);
    exit;
  end;

  try
    qry := Price.ListPrice('', erro);
    if qry.RecordCount > 0 then
    begin
      objPrice := qry.ToJSONObject;
      Res.Send<TJSONObject>(objPrice)
    end
    else
    begin
      Res.Send('Produto não localizado. Verifique os filtros utilizados')
        .Status(404);
    end;
  finally
    qry.Free;
    Price.Free;
  end;
end;

{ procedure consume endpoint }
procedure Endpoint;
begin
  { declare http uses }
  THorse.Get('/price', ListPrice);
  THorse.Get('/price/:codauxiliar/:codfilial/:numregiao', ListPrice);

end;

end.

`Model.Price`

unit Model.Price;

interface

uses FireDAC.Comp.Client, Data.DB, System.SysUtils, Model.Connection;

type
  TPrice = class

  private
    //FPRECOOFERTA: String;
    FCODPROD: String;
    FQTMINIMAATACADO: String;
    FPVENDAATAC: String;
    FPVENDA: String;
    FCODAUXILIAR: String;
    FCODFILIAL: String;
    FNUMREGIAO: String;

  public
    constructor Create;
    destructor Destroy; override;
    property CODPROD: String read FCODPROD write FCODPROD;
    property PVENDA: String read FPVENDA write FPVENDA;
    property PVENDAATAC: String read FPVENDAATAC write FPVENDAATAC;
    property QTMINIMAATACADO: String read FQTMINIMAATACADO write FQTMINIMAATACADO;
   // property PRECOOFERTA: String read FPRECOOFERTA write FPRECOOFERTA;
    property CODAUXILIAR: String read FCODAUXILIAR write FCODAUXILIAR;
    property CODFILIAL: String read FCODFILIAL write FCODFILIAL;
    property NUMREGIAO: String read FNUMREGIAO write FNUMREGIAO;
    function ListPrice(OrderBy: String; Out Error: String): TFDQuery;

  end;

implementation

{ TPrice }

constructor TPrice.Create;
begin
  Model.Connection.Connect;
end;

destructor TPrice.Destroy;
begin
  Model.Connection.Disconect;
  inherited;
end;

function TPrice.ListPrice(OrderBy: String; out Error: String): TFDQuery;
var
  qry: TFDQuery;
begin
  try
    qry := TFDQuery.Create(nil);
    qry.Connection := Model.Connection.FConnection;
    with qry do
    begin
      Active := False;
      SQL.Clear;
      SQL.Add(
      'select e.codprod, '
      + 'nvl(pctabpr.pvenda, 0) * decode(nvl(e.fatorpreco, 1), 0, 1, nvl(e.fatorpreco, 1)) * decode(nvl(e.qtunit, 1), 0, 1, nvl(e.qtunit, 1)) pvenda, '
      + 'nvl(pctabpr.pvendaatac, 0) * decode(nvl(e.fatorpreco, 1), 0, 1, nvl(e.fatorpreco, 1)) * decode(nvl(e.qtunit, 1), 0, 1, nvl(e.qtunit, 1)) pvendaatac, '
      + 'nvl((select pcprodfilial.qtminimaatacado from pcprodfilial where codprod = pctabpr.codprod and numregiao = pctabpr.numregiao '
      + 'and codfilial = e.codfilial), e.qtminimaatacado) qtminimaatacado '
      + 'from pctabpr, pcembalagem e '
      + 'where 1=1 '
      + 'and pctabpr.codprod = e.codprod '
      );

      if CODAUXILIAR <> '' then
        begin
          SQL.Add('and e.codauxiliar = :codauxiliar');
          ParamByName('CODAUXILIAR').Value := CODAUXILIAR;
        end;

      if CODFILIAL <> '' then
        begin
          SQL.Add('and e.codfilial = :codfilial');
          ParamByName('CODFILIAL').Value := CODFILIAL;
        end;

         if NUMREGIAO <> '' then
        begin
          SQL.Add('and pctabpr.numregiao = :numregiao');
          ParamByName('NUMREGIAO').Value := NUMREGIAO;
        end;

      Active := True;
    end;
    Error := '';
    Result := qry;
  except
    on Ex: Exception do
    begin
      Error := 'Erro ao realizar a consulta ao banco de dados' + #13 + 'Erro: '
        + Ex.Message;
      Result := nil;
    end;
  end;
end;

end.

**Comment by viniciussanchez** [comment, marked as answer]
Da uma olhada nesses vídeos:
https://youtu.be/Cm78uY34OJI
https://youtu.be/haI4-DazVXg
https://youtu.be/9HNqjpjc5XA

**Reply by alvaroflp** [reply]
Valeu demais, foi de grande valia os vídeos. Consegui resolver revisando o código e vi que estava chamando a função ListPrice ao invés de ListPriceParams.

**Reply by viniciussanchez** [reply]
Legal... que bom que ajudou
